src/Authentication/Authentication/ManagedServiceAccessTokenWebAppsMsiAdHoc.cs

```diff
@@ -0,0 +1,132 @@
+﻿// ----------------------------------------------------------------------------------
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using Microsoft.Azure.Commands.Common.Authentication.Properties;
+using Microsoft.Rest.Azure;
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+
+namespace Microsoft.Azure.Commands.Common.Authentication
+{
+    public class ManagedServiceAccessTokenWebAppsMsiAdHoc : IRenewableToken
+    {
+        private readonly IAzureAccount _account;
+        private const string ResourceId = @"https://management.azure.com/";
+        private readonly IHttpOperations<ManagedServiceTokenInfoWebAppsMsiAdHoc> _tokenGetter;
+        private DateTime _expiration = DateTime.UtcNow;
+        private string _accessToken;
+
+        public ManagedServiceAccessTokenWebAppsMsiAdHoc(IAzureAccount account, IAzureEnvironment environment, string tenant = "Common")
```

Comment:
It seems like, if we create a common interface between the MSI payload types, we can reuse the existing ManagedServiceAccessToken - it would be nice to reuse the code - whether that is by adding a new constructor or introducing an abstract base class.